src/dispatch/RetryHttpHandler.ts

```diff
@@ -0,0 +1,41 @@
+import { HttpHandler, HttpRequest, HttpResponse } from '@aws-sdk/protocol-http';
+
+/**
+ * An HttpHandler which wraps other HttpHandlers to retry requests.
+ *
+ * Requests will be retried if (1) there is an error (e.g., with the network or
+ * credentials) and the promise rejects, or (2) the response status is not 2xx.
+ */
+export class RetryHttpHandler implements HttpHandler {
+    private handler: HttpHandler;
+    private retries: number;
+
+    public constructor(handler: HttpHandler, retries: number) {
+        this.handler = handler;
+        this.retries = retries;
+    }
+
+    public async handle(
+        request: HttpRequest
+    ): Promise<{ response: HttpResponse }> {
+        let retriesLeft = this.retries;
+        while (true) {
+            try {
+                const response = await this.handler.handle(request);
+                if (this.isStatusCode2xx(response.response.statusCode)) {
```

Comment:
issue: I think we should perhaps treat 4xx and 5xx from data plane separately - if it's 5xx, we know it's a service issue, and hopefully, the issue will be mitigated fairly quickly. It's something we can fix on our end, so once things are recovered on our end, the web client should be able to send requests from there on. On the other hand, if it's 4xx, most likely the setup or credentials are invalid, and without the customer's own fixes, there's no point in sending more events (which is the whole purpose of adding this feature).